### PR TITLE
fix: update streamer marker validation test

### PR DESCRIPTION
**Pull request by narenbagria**
Resolves: [VLCLJ-2344](https://jira.devtools.intel.com/browse/VLCLJ-2344)

### CODE REVIEW

File: conformance_tests/tools/metrics/src/test_metric.cpp [hunk 1]
```diff
@@ -2063,29 +2078,53 @@ TEST_F(
...
+          maxDuration = std::min(maxDuration, 100u);
+        }
+        auto start_time = std::chrono::high_resolution_clock::now();
+        auto duration = std::chrono::milliseconds(maxDuration);
+
+        while ((std::chrono::high_resolution_clock::now() - start_time) <
+               duration) {
+          std::vector<uint8_t> fetchedData{};
+          uint32_t fetchedDataSize{};
+          lzt::metric_streamer_read_data(metricStreamerHandle, fetchedDataSize,
```
Comment (joshuaranjan) [inline conformance_tests/tools/metrics/src/test_metric.cpp:2098]: Cant we have a short sleep 1 ms sleep before we read the data every time ?

File: utils/test_harness/tools/src/test_harness_metric.cpp [hunk 2]
```diff
@@ -1022,22 +1022,24 @@ void metric_validate_streamer_marker_data(
             (metricTypedValue.value.ui64 != 0)) {
           LOG_INFO << "Valid Streamer Marker found with value: "
                    << metricTypedValue.value.ui64;
-          EXPECT_EQ(metricTypedValue.value.ui64,
-                    streamerMarkerValues[streamerMarkerValuesIndex]);
+          EXPECT_EQ(streamerMarkerValues[streamerMarkerValuesIndex],
```
Comment (joshuaranjan) [inline utils/test_harness/tools/src/test_harness_metric.cpp:1025]: Why is this changed ?

File: conformance_tests/tools/metrics/src/test_metric.cpp [hunk 3]
```diff
@@ -2063,29 +2078,53 @@ TEST_F(
...
+          lzt::metric_get_metric_handles_from_metric_group(
+              groupInfo.metricGroupHandle, metricHandles);
+          std::vector<zet_metric_properties_t> metricProperties(
+              metricHandles.size());
+          lzt::metric_get_metric_properties_for_metric_group(metricHandles,
+                                                             metricProperties);
+
+          lzt::metric_validate_streamer_marker_data(
+              metricProperties, metricValues, metricValueSets,
+              streamerMarkerValues, numMarkersFound);
```
Comment (joshuaranjan) [inline conformance_tests/tools/metrics/src/test_metric.cpp:2121]: Are we removing the markers which were already found from streamerMarkerValues ?

File: conformance_tests/tools/metrics/src/test_metric.cpp [hunk 4]
```diff
@@ -2063,29 +2078,53 @@ TEST_F(
...
+
+        while ((std::chrono::high_resolution_clock::now() - start_time) <
+               duration) {
+          std::vector<uint8_t> fetchedData{};
+          uint32_t fetchedDataSize{};
+          lzt::metric_streamer_read_data(metricStreamerHandle, fetchedDataSize,
+                                         &fetchedData);
+
+          fetchedData.resize(fetchedDataSize);
+          rawData.insert(rawData.end(), fetchedData.begin(), fetchedData.end());
```
[the remark this answers is not in the document]
Comment (matcabral) [reply conformance_tests/tools/metrics/src/test_metric.cpp:2102]: >I think concatenation would not work for calculate on a root device with multiple sub-devices. 

I think this is supported in. Please test in PVC

However, I'm not sure of the logic followed. It seems the loop is reading more data on each iteration and concatenating to existing data and calling calculate on the concatenated data on every iteration. This is not very efficient, since it recalculates the same data on every iteration. Why not just call calculate on the new data of each loop ?

File: conformance_tests/tools/metrics/src/test_metric.cpp [hunk 5]
```diff
@@ -2008,6 +2022,7 @@ TEST_F(
       LOG_INFO << "test metricGroup name " << groupInfo.metricGroupName;
 
       std::vector<uint32_t> streamerMarkerValues = {10, 20};
+      uint32_t numMarkersFound{};
```
[the remark this answers is not in the document]
Comment (narenbagria) [reply conformance_tests/tools/metrics/src/test_metric.cpp]: Resolved, Thank you.

File: conformance_tests/tools/metrics/src/test_metric.cpp [hunk 6]
```diff
@@ -2063,29 +2078,53 @@ TEST_F(
...
-                                                           metricProperties);
-
-        lzt::metric_validate_streamer_marker_data(metricProperties,
-                                                  metricValues, metricValueSets,
-                                                  streamerMarkerValues);
+        std::vector<uint8_t> rawData{};
+        uint32_t rawDataSize{};
+        const char *valueString =
+            std::getenv("LZT_METRIC_READ_DATA_MAX_DURATION_MS");
+        uint32_t maxDuration = 5; // millisecond(s)
```
[the remark this answers is not in the document]
Comment (narenbagria) [reply conformance_tests/tools/metrics/src/test_metric.cpp]: Resolved, Thank you.

File: conformance_tests/tools/metrics/src/test_metric.cpp [hunk 7]
```diff
@@ -2063,29 +2078,53 @@ TEST_F(
...
+          maxDuration = value != 0 ? value : maxDuration;
+          maxDuration = std::min(maxDuration, 100u);
+        }
+        auto start_time = std::chrono::high_resolution_clock::now();
+        auto duration = std::chrono::milliseconds(maxDuration);
+
+        while ((std::chrono::high_resolution_clock::now() - start_time) <
+               duration) {
+          std::vector<uint8_t> fetchedData{};
+          uint32_t fetchedDataSize{};
```
[the remark this answers is not in the document]
Comment (narenbagria) [reply conformance_tests/tools/metrics/src/test_metric.cpp]: Resolved, Thank you.

File: conformance_tests/tools/metrics/src/test_metric.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (narenbagria) [reply conformance_tests/tools/metrics/src/test_metric.cpp]: Updated the implementation, Thank you.

File: conformance_tests/tools/metrics/src/test_metric.cpp (hunk 1 above)
Comment (narenbagria) [reply conformance_tests/tools/metrics/src/test_metric.cpp]: Resolved, Thank you.

File: utils/test_harness/tools/src/test_harness_metric.cpp (hunk 2 above)
Comment (narenbagria) [reply utils/test_harness/tools/src/test_harness_metric.cpp]: It's recommended to have actual value as first argument and expected value as second, Thank you.

File: conformance_tests/tools/metrics/src/test_metric.cpp [hunk 8]
```diff
@@ -2063,29 +2077,51 @@ TEST_F(
...
+        while ((std::chrono::high_resolution_clock::now() - start_time) <
+                   duration &&
+               (streamer_marker_values_index < streamerMarkerValues.size())) {
+          std::vector<uint8_t> rawData{};
+          uint32_t rawDataSize{};
+          std::this_thread::sleep_for(std::chrono::milliseconds(1));
+          lzt::metric_streamer_read_data(metricStreamerHandle, rawDataSize,
+                                         &rawData);
+
+          rawData.resize(rawDataSize);
```
Comment (joshuaranjan) [inline conformance_tests/tools/metrics/src/test_metric.cpp:2102]: can we resize after the "if" condition ?

File: conformance_tests/tools/metrics/src/test_metric.cpp [hunk 9]
```diff
@@ -2063,29 +2078,53 @@ TEST_F(
...
-        lzt::metric_get_metric_properties_for_metric_group(metricHandles,
-                                                           metricProperties);
-
-        lzt::metric_validate_streamer_marker_data(metricProperties,
-                                                  metricValues, metricValueSets,
-                                                  streamerMarkerValues);
+        std::vector<uint8_t> rawData{};
+        uint32_t rawDataSize{};
+        const char *valueString =
+            std::getenv("LZT_METRIC_READ_DATA_MAX_DURATION_MS");
```
[the remark this answers is not in the document]
Comment (narenbagria) [reply conformance_tests/tools/metrics/src/test_metric.cpp:2081]: I'm planning to address all the macros at once, similar to how it will be done for the embargo [VLCLJ-2279](https://jira.devtools.intel.com/browse/VLCLJ-2279). Thank you.

File: conformance_tests/tools/metrics/src/test_metric.cpp [hunk 10]
```diff
@@ -2063,29 +2077,51 @@ TEST_F(
...
+          uint32_t value = atoi(value_string);
+          max_duration_in_milliseconds =
+              value != 0 ? value : max_duration_in_milliseconds;
+          max_duration_in_milliseconds =
+              std::min(max_duration_in_milliseconds, 100u);
+        }
+        auto start_time = std::chrono::high_resolution_clock::now();
+        auto duration = std::chrono::milliseconds(max_duration_in_milliseconds);
+        uint32_t streamer_marker_values_index = 0;
+        while ((std::chrono::high_resolution_clock::now() - start_time) <
```
Comment (joshuaranjan) [inline conformance_tests/tools/metrics/src/test_metric.cpp:2093]: Please add a comment here about the loop.
Something like "Run till ..."

File: conformance_tests/tools/metrics/src/test_metric.cpp [hunk 11]
```diff
@@ -2063,29 +2077,51 @@ TEST_F(
...
+                   duration &&
+               (streamer_marker_values_index < streamerMarkerValues.size())) {
+          std::vector<uint8_t> rawData{};
+          uint32_t rawDataSize{};
+          std::this_thread::sleep_for(std::chrono::milliseconds(1));
+          lzt::metric_streamer_read_data(metricStreamerHandle, rawDataSize,
+                                         &rawData);
+
+          if (rawDataSize == 0) {
```
Comment (joshuaranjan) [inline conformance_tests/tools/metrics/src/test_metric.cpp:2102]: Please add a comment here

File: conformance_tests/tools/metrics/src/test_metric.cpp [hunk 12]
```diff
@@ -2063,29 +2077,51 @@ TEST_F(
...
+          std::vector<zet_metric_properties_t> metricProperties(
+              metricHandles.size());
+          lzt::metric_get_metric_properties_for_metric_group(metricHandles,
+                                                             metricProperties);
+
+          lzt::metric_validate_streamer_marker_data(
+              metricProperties, metricValues, metricValueSets,
+              streamerMarkerValues, streamer_marker_values_index);
+        }
+        EXPECT_EQ(streamerMarkerValues.size(), streamer_marker_values_index);
```
Comment (joshuaranjan) [inline conformance_tests/tools/metrics/src/test_metric.cpp:2124]: Please add a comment about this EXPECT.
"Checking if .."

File: conformance_tests/tools/metrics/src/test_metric.cpp [hunk 13]
```diff
@@ -2063,29 +2077,51 @@ TEST_F(
...
+          lzt::metric_calculate_metric_values_from_raw_data(
+              groupInfo.metricGroupHandle, rawData, metricValues,
+              metricValueSets);
+
+          std::vector<zet_metric_handle_t> metricHandles;
+          lzt::metric_get_metric_handles_from_metric_group(
+              groupInfo.metricGroupHandle, metricHandles);
+          std::vector<zet_metric_properties_t> metricProperties(
+              metricHandles.size());
+          lzt::metric_get_metric_properties_for_metric_group(metricHandles,
```
Comment (joshuaranjan) [inline conformance_tests/tools/metrics/src/test_metric.cpp:2117]: Could we move this outside the "repeated-reading" loop ?

File: conformance_tests/tools/metrics/src/test_metric.cpp [hunk 14]
```diff
@@ -2063,29 +2077,51 @@ TEST_F(
...
-            metricHandles.size());
-        lzt::metric_get_metric_properties_for_metric_group(metricHandles,
-                                                           metricProperties);
-
-        lzt::metric_validate_streamer_marker_data(metricProperties,
-                                                  metricValues, metricValueSets,
-                                                  streamerMarkerValues);
+        const char *value_string =
+            std::getenv("LZT_METRIC_READ_DATA_MAX_DURATION_MS");
+        uint32_t max_duration_in_milliseconds = 10;
```
Comment (joshuaranjan) [inline conformance_tests/tools/metrics/src/test_metric.cpp:2082]: Please change max_duration_in_milliseconds -> max_wait_time_in_milliseconds, if it makes sense

**Review by joshuaranjan — COMMENTED** [review]
functionally looks Ok.
Please check the feedback about improving readability.

File: conformance_tests/tools/metrics/src/test_metric.cpp (hunk 10 above)
Comment (narenbagria) [reply conformance_tests/tools/metrics/src/test_metric.cpp:2108]: Resolved. Thank you.

File: conformance_tests/tools/metrics/src/test_metric.cpp (hunk 12 above)
Comment (narenbagria) [reply conformance_tests/tools/metrics/src/test_metric.cpp:2133]: Resolved. Thank you.

**Review by matcabral — APPROVED** [review]

**Review by joshuaranjan — APPROVED** [review]

**Review by vishnu-khanth — APPROVED** [review]

**Review by shubskmr — APPROVED** [review]
